docs/proposals/dynamic-kubelet-settings.md

```diff
@@ -0,0 +1,129 @@
+- `<node-name>-kubelet-configuration`: the `KubeletConfiguration` configmap specific to the node named `<node-name>`, also set via the API server.
+
+### Well-defined configmap type for Kubelet configuration
+
+- There is an alpha version of the `KubeletConfiguration` type: external in `pkg/componentconfig/v1alpha1/types.go`, internal in `pkg/componentconfig/types.go`.
+- We likely want to group related fields of the `KubeletConfiguration` into substructures. Please note that `v1alpha1` will not be regrouped, but eventually we will start working on this.
+
+@timstclair and I did some brainstorming and came up with some rough potential categories:
+
+---
+- Networking
+- Runtimes (only one of these, in addition to Common, should be specified for a given Kubelet)
+    + Common
+    + Rkt
+    + Docker
+- Events
+- Security/Policies
+- Images (might be better as a runtime sub-category)
+- Resource Management
+    + GC/Eviction
+    + Allocatable/SystemReserved/KubeReserved
+    + Custom resources? e.g. NvidiaGPUs?
+- Cloud-provider-specific settings
+- Cluster/API
+    + KubeAPIQPS
+    + ContentType
+- SystemSpecification (this might be a non-category, all of the available hardware should theoretically be detectable by the Kubelet)
+    + NodeIP (probably belongs in networking)
+    + NvidiaGPUs could be here, but shouldn't the Kubelet be able to detect this?
+- Experimental/Volatile Features (could be tricky, might be better off using alpha versioning to separate experiments from the stable API)
+
+---
+
+### Bringing up a master node
+
+- Start the Kubelet with `minimal-config`
+- On-disk static pod manifests are used to bring up the API server, scheduler, controller-manager, etc.
+Once the master is ready, the user creates some configs via the API:
+- They might create a `global-kubelet-configuration`
+- They might create a `<node-name>-kubelet-configuration`, e.g. if they want a master-specific configuration.
+- Then the Kubelets' configuration sync loops will see the new config, and the Kubelets will restart to take up that new config.
+
+### Bringing up a Kubelet
+
+- Start the Kubelet with `minimal-config`
+- If `<current-node-name>-kubelet-config` exists on the API server, download and use the `<current-node-name>-kubelet-config`
+- else if `global-kubelet-config` exists on the API server, download and use the `global-kubelet-config`.
+- else if e.g. an `on-disk-config` flag is set, defer to the node's on-disk config.
+- else defer to config from flags
+- Finally, start the configuration sync loop in a background thread
+
+
+### Storing config locally (checkpointing)
+
+Initially, the Kubelet will store the most recent configmap it got from the API server internally in memory, and compare that to new values it gets from the API server. It will restart when these configmaps differ in order to apply the new config.
```

Comment:
Sorry, I was misunderstanding the process here. So the kubelet will just see a diff of configs, throw it out, restart, pull down the new config and apply it early-ish in its initialization.